evm/auxiliary/user_account_contract/contract.py

```diff
@@ -0,0 +1,211 @@
+NONCE_GETTER_ID = big_endian_to_int(keccak(b"get_nonce()")[:4])
+VALIDATION_CODE_GETTER_ID = big_endian_to_int(keccak(b"get_validation_code()")[:4])
+ENTRY_POINT_INT = big_endian_to_int(ENTRY_POINT)
+
+SIGNATURE_VERIFICATION_ADDRESS = 1
+SIGNATURE_VERIFICATION_GAS = 3000
+
+# calldata locations
+CALLDATA_FUNCTION_ID = 0
+CALLDATA_SIGNATURE = 0
+(
+    CALLDATA_V,
+    CALLDATA_R,
+    CALLDATA_S,
+    CALLDATA_NONCE,
+    CALLDATA_GASPRICE,
+    CALLDATA_VALUE,
+    CALLDATA_MIN_BLOCK,
+    CALLDATA_MAX_BLOCK,
+    CALLDATA_DESTINATION,
+    CALLDATA_DATA,
+) = (CALLDATA_SIGNATURE + i * 32 for i in range(10))
+
+# memory locations
+MEMORY_CALLDATA = 32
+MEMORY_NONCE = MEMORY_CALLDATA + CALLDATA_NONCE
+MEMORY_DATA = MEMORY_CALLDATA + CALLDATA_DATA
+MEMORY_SIGHASH = 0
+MEMORY_SENDER = 0
+MEMORY_RETURNCODE = 0
+
+# storage locations
+STORAGE_NONCE = 0
+
+
+def generate_account_lll_code(validation_code_address):
+    return [
+        'seq',
+
+        # handle get_nonce if called
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                NONCE_GETTER_ID],
+            [
+                'seq',
+                ['mstore', 0, ['sload', STORAGE_NONCE]],
+                ['return', 0, 32]
+            ]],
+
+        # copy full call data to memory
+        ['calldatacopy', MEMORY_CALLDATA, 0, ['calldatasize']],
+
+        # handle get_validation_code if called
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                VALIDATION_CODE_GETTER_ID,
+                # return validation code address
+                ['mstore', 0, validation_code_address],
+                ['return', 12, 20]]],
+
+        # no function is called, so we should be first in the call stack
+        ['assert', ['eq', ['caller'], ENTRY_POINT_INT]],
+
+        # check if block is in valid range
+        ['assert', ['ge', ['number'], ['calldataload', CALLDATA_MIN_BLOCK]]],
+        ['assert', ['le', ['number'], ['calldataload', CALLDATA_MAX_BLOCK]]],
+
+        # Check for small s to avoid malleability
+        ['assert', ['lt', ['calldataload', CALLDATA_S], MAX_ALLOWED_S + 1]],
+
+        # load tx-sighash to the end of memory
+        ['mstore', ['add', MEMORY_CALLDATA, ['calldatasize']], ['sighash']],
+        # Compute sighash = sha3(nonce ++ gasprice ++ value ++ min-block ++ max-block ++ to ++
+        # data ++ tx-sighash)
+        ['mstore',
+            MEMORY_SIGHASH,
+            ['sha3', MEMORY_NONCE, ['sub', ['calldatasize'], CALLDATA_NONCE - 32]]],
+
+        # Verify signature by calling validation code
+        ['assert',
+            [
+                'call',
+                VALIDATION_CODE_GAS,
+                validation_code_address,
+                0,
+                # input data (spans sighash and signature)
+                MEMORY_SIGHASH,
+                128,
+                # output data (discarded)
+                0,
+                0
+            ]],
```

Comment:
Let's not do things this way. This would require calling 2 accounts to send from this account, which would blow up witness size by a factor of 2. IMO just do ecrecover here.